### ISSUE TITLE
Mega Issue: Manual node provisioning

### ISSUE THREAD

**Issue by ellistarn** [aws/karpenter-core#749]
**Tell us about your request**
What do you want us to build?

I'm seeing a number of feature requests to launch nodes separately from pending pods. This issue is intended to broadly track this discussion. 

Use Cases:
- Create a System pool to run components like karpenter, loadbalancer, coredns, etc
- Provision baseline capacity that never scales down
- Manually preprovision a set of nodes before a large event

### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

**Comment by chamarakera** [comment]
When migrating from Cluster Auto Scaler to Karpenter, we would like Karpenter to provision a node beforehand, before we perform a drain on the old node. It takes time for Karpenter to provision a new node based on the unscheduled workloads, and due to this, the pods are kept in a Pending state for too long.

**Comment by Bryce-Soghigian** [comment]
@chamarakera would just requesting a node via nodeclaim be enough in this case? No managed solution is requried if you just want to create a node from my testing just applying nodeclaims with a reference to valid nodepool is often enough. 


Note this example uses a instance type size from azure. 

k apply -f nodeclaim.yaml

```yaml
apiVersion: karpenter.sh/v1beta1
kind: NodeClaim
metadata:
  name: temporary-capacity
  labels:
    karpenter.sh/nodepool: general-purpose
  annotations:
    karpenter.sh/do-not-disrupt: "true"
spec:
  nodeClassRef:
    name: default
  requirements:
  - key: kubernetes.io/arch
    operator: In
    values:
    - amd64
  - key: kubernetes.io/os
    operator: In
    values:
    - linux
  - key: karpenter.sh/capacity-type
    operator: In
    values:
    - on-demand
  - key: node.kubernetes.io/instance-type
    operator: In
    values:
    - Standard_D32_v3
  resources:
    requests:
      cpu: 2310m
      memory: 725280Ki
      pods: "7"
status: {}
 ```

**Comment by njtran** [comment]
@chamarakera how bad is the pod latency you're describing here? Do you have a long bootstrap/startup time on your instance? What would be an acceptable amount of pod latency to not need prewarmed instances ?

**Comment by chamarakera** [comment]
@Bryce-Soghigian - To use NodeClaims, we would need to generate several NodeClaim resources depending on the cluster's size. I would like to see a configurable option within the NodePool itself to provision nodes prior to the migration. I think, having a way to configure min size/max size parameters in the NodePool itself would be a good solution.

@njtran - Usually it takes for 1 - 2mins for startup, this is ok in non-prod,  but in a production I would like to have all pods scheduled in a node as soon as possible (few seconds).

**Comment by garvinp-stripe** [comment]
After chatting a bit in Slack on the Karpenter channel with awiesner4 I think i have some thoughts around this problem. 

First I want to contrast Karpenter's primary objective then break down Karpenter's current responsibility and maybe this will help drive the design choice. I think Karpenter's main objective is be an efficiency cluster autoscaler so it makes it difficult for leaving around nodes that isn't doing work go against what Karpenter is trying to achieve. I think to add something that works around it would likely be problematic because you would have to work around everything Karpenter is built to do. 

However, at the moment is doing more than just autoscaling which is where I think problems and usability issue arises. It autoscales but it also manages nodes. It takes over how nodes are configured and the lifecycle of nodes and it closes the door for other things to manage nodes. 

https://github.com/kubernetes-sigs/karpenter/issues/742
https://github.com/kubernetes-sigs/karpenter/issues/688
https://github.com/kubernetes-sigs/karpenter/issues/740
+ this issue.

What does this mean? I agree with those who are saying nodeclaim should be able to create nodes essentially outside of Karpenter's main autoscaling logic so we don't change what Karpenter is trying to do (save money). I think at this time NodePools does hold the logic and concept of how Karpenter tries to optimize the cluster so I don't think manual provisioning should live there. On the max/min nodes on NodePool, it was pointed out to me that how would a nodepool know what instance types those min nodes should be and in order to protect those min nodes Karpenter would have to cut through most of its disruption logic to support _don't drop node count below min_. 

That isn't the say supporting node management or different autoscaling priority isn't possible but I think the entity that contains that logic should not be node pool in its current form. If Karpenter expands NodePool such that it is extensible, _karpenter_nodepool_provider_N, that allows users to group nodeclaims with different intention that differs from the primary objective of Karpenter users can create nodepools variants where keeping a min make sense. Where scheduling logic is different and so on.

From my view, I think its important to keep Karpenter's main focus clear and clean because its complicated enough. But if we allow for more extension on the base concepts, we may be able to support use cases to fall out of what Karpenter is trying to do.

**Comment by Bryce-Soghigian** [comment]
> However, at the moment is doing more than just autoscaling which is where I think problems and usability issue arises.

There are mentions of node autohealing, using budgets to manage upgrades, etc. Seems its moving to be a node lifecycle management tool. It has more value than just statically autoscaling. So I am for a static provisioning CR that helps manage lifecycle of static pools.

**Comment by SatishReddyBethi** [comment]
Hi. I am really looking forward to this feature too. Is there an open PR for it? Or is it still in the discussions phase?

**Comment by cloudwitch** [comment]
We have 9 minimum nodes in our ASG for a batch job workload that gets kicked off by users through a UI.

The users find the EC2 spin-up time unacceptable and expect their pod to spin up quicker than the EC2 can start.

We must have a way to run a minimum number of nodes in a nodepool.

**Comment by sftim** [comment]
> We must have a way to run a minimum number of nodes in a nodepool.

You can already do that (run low-priority placeholder Pods), but AFAIK there's no controller that does exactly this.
Maybe I'll put some time in and try to write one.

**Comment by tallaxes** [comment]
Cluster Autoscaler now also supports [ProvisioningRequest CRD](https://github.com/kubernetes/autoscaler/blob/master/cluster-autoscaler/proposals/provisioning-request.md)

**Comment by evanlivingston** [comment]
There is an additional financial impact caused by the de-facto solution for a warm pool of capacity by overprovisioning. When high priority pods are scheduled, preempting the `overprovision` workloads, Karpenter will scale the nodegroup immediately in order to regain capacity to reschedule the `overprovision` workloads. In some cases it may be preferable for the headroom to be elastic, such that the desired headroom is restored after a deployment or batch job completes. This could be accomplished by setting a desired _range_ of overprovisioned resources. But I also believe @sftim `minimumPodPriority` solution presents an acceptable solution.

**Comment by riiv-hexagon** [comment]
+1

**Comment by ellistarn** [comment]
Cross ref-ing the ProvisioningRequest ask here: https://github.com/kubernetes-sigs/karpenter/issues/742#issuecomment-2122005473

**Comment by jackfrancis** [comment]
cc @raywainman who is tracking warm replicas stories on behalf of WG Serving here: 

https://docs.google.com/document/d/1QsN4ubjerEqo5L4bQamOFFS2lmCv5zNPis2Z8gcIITg

**Comment by jonathan-innis** [comment]
For everyone's context, we did a little bit of ideating and came-up with an API that we were pretty happy with from the Karpenter side (see https://github.com/jonathan-innis/karpenter-headroom-poc). We're having an open discussion with the CAS folks about the difference between how we are thinking about the Headroom API and the ProvisioningRequest API, feel free to take a look and comment on the doc if you have any thoughts: https://docs.google.com/document/d/1SyqStWUt407Rcwdtv25yG6MpHdNnbfB3KPmc4zQuz1M/edit?usp=sharing

**Comment by balmha** [comment]
Hi guys, how's it going?. I have been looking for a fixed provisioning solution with Karpenter-

We need to have a minimum amount of on-demand nodes (I'd say 2, one per AZ) and then scale-up/down with spot instances the rest of the workload, I haven't found a solution yet, we tried to play with "On-Demand/Spot Ratio Split" but didn't work and continues spreading Spot instances for the whole workload.

Some workaround or thougts about how to solve it, we really want to fully use Karpenter for our workload.

**Comment by gazal-k** [comment]
> We need to have a minimum amount of on-demand nodes (I'd say 2, one per AZ) and then scale-up/down with spot instances the rest of the workload, I haven't found a solution yet, we tried to play with "On-Demand/Spot Ratio Split" but didn't work and continues spreading Spot instances for the whole workload.

You don't need this issue resolved for your use case @balmha.
- Create 1 `NodePool` with requirement `karpenter.sh/capacity-type` set to `["on-demand"]` and higher `weight`. Set `limits` to control how much on-demand "base capacity" you need. 
- Create another `NodePool` with requirement `karpenter.sh/capacity-type` set to `["spot", "on-demand"]` (it'll schedule spot nodes and only fallback to on-demand in case of availability issues) with lower `weight`.

Even better, you can let your workloads that can't tolerate interruptions, set nodeSelectors or affinity to run on on-demand nodes. That's something which is cleaner to do with Karpenter than on CA.

**Comment by stevehipwell** [comment]
@jonathan-innis I really like the look of the headroom APIs, I think they'd cover the requirements I was talking about in #993. Is there a separate issue to track the headroom APIs? Google docs are blocked from our corporate machines.

**Comment by Scrat94** [comment]
Is this still planned to be implemented? Any kind of ETA? 

Our use case also would greatly benefit from it: We scale up GitLab Runners for our organization. However the cold-starts (60-90 secs) is not ideal for CI/CD. Having 1-2 nodes always available as headroom would make sure that every pipeline coming in, directly get started. Ideally, the minimum number of nodes can be set for a specific schedule (e.g. office hours only) to make sure that during night or weekend it ramps down to zero to minimize costs. 
(Or is there any other approach I can choose to fulfill my usecase with e.g. EKS + Karpenter?)

**Comment by viettranquoc** [comment]
Our team is using blue/green deployment and always needs to wait a few minutes to make the new nodes come. And then Karpenter will disrupt nodes (after deployment, there will be an overuse of resources). 
It would be great if we could allow a buffer like CPU or Memory. For example, if we constantly use a 1000m CPU, it would be a 1200m—20% buffer for deployment and cronjob running.

**Comment by woehrl01** [comment]
One suggestion for always have a fixed set of nodes available is heaving a deployment with proper topologyspreadconstraints to force the pods one node per replica. (set very low requirements) 

Don't set a lower priority so it won't get prempted (as in the prewarm/headroom Szenario mentioned before). 

If you combine this with keda you can do that even dynamically basted on external factors.

**Comment by gmoirod** [comment]
Hi there !

Just wanted to share how I manage to get something working for me.

My need is for a fleet of CI runners : they are spawned for each job in my pipeline and waiting for new nodes is tooslow.
What I did :

    create a deployment "warmup" with a K8S low PriorityClass and requiring no other pod on same node (by anti-affinity)
    define resources.requests big enough to occupy a node alone
    for other deployments/pods (here my CI runners) : give a higher K8S PriorityClass + a PDB to forbid Karpenter to evict them
    allow Karpenter to consolidate at any moment (to delete unused nodes after pipeline execution)

This way, when a CI job needs resources, it evicts the "warmup" pod to take the node. Then the "warmup" pod claim a new spare node.

This works pretty well for me :

    I can control how many "warm up" nodes I want with the replicaCount
    I use Spot instances so my CI nodepool is cost effective
    Pipelines are quite fast. Most of the time they do not wait for new nodes
    Karpenter cleans all unused nodes quickly

If it can help before a feature allow that ;)

**Comment by sftim** [comment]
https://github.com/kubernetes-sigs/karpenter/issues/749#issuecomment-2607044025 sounds similar to https://kubernetes.io/docs/tasks/administer-cluster/node-overprovisioning/

However, there are still end users who'd like more direct control over NodeClaims and the nodes that Karpenter provisions to satisfy those claims. Also, people who want a minimum resource capacity in a NodePool even when no pods match it.

**Comment by morgayle-ngt** [comment]
Hello @ellistarn! 🙌 My comments above are simply meant to:

- Acknowledge the community’s valuable insights--your use-case breakdown was especially helpful!

- Suggest actionable next steps to unblock users today while iterating toward a robust solution

- Seek alignment with Karpenter maintainers on design direction, ensuring proposals align with the project’s long-term vision.

Appreciate everyone’s thoughtful contributions here! 🚀


**Immediate Workaround:** 

- Recommend the dual NodePool strategy (on-demand base + spot/on-demand fallback) as a current best practice. We can leverage Karpenter's existing weight/limit system and pod affinity rules to maintain baseline capacity without requiring code changes.
- Solves CI/CD "cold start" delays by maintaining a warm on-demand pool for burstable workloads (e.g., pipeline jobs).
- Avoids timeouts/failures from waiting for spot node provisioning during pipeline executions.

**Short-Term Enhancement:** 

- Propose implementing maxSurge/maxUnavailable semantics in NodePool CRD for controlled scaling operations. This would address rolling updates and HA requirements while aligning with Kubernetes patterns.
- Ensures nodes are pre-provisioned before old nodes are terminated during rolling updates, preventing CI/CD disruptions during deployments.
- Aligns with Kubernetes Deployment patterns, giving CI/CD systems predictable scaling behavior.

**Long-Term Roadmap:** 

- Consider introducing explicit CapacityReservation or WarmPool APIs in future Karpenter versions for first-class cold start mitigation, potentially building on the existing disruption mechanisms.
- Explicitly reserves capacity for critical workloads like CI/CD runners, eliminating cold starts entirely.
- Allows CI/CD systems to "pre-warm" nodes before large jobs (e.g., parallel test runs).

**CI/CD Impact:** 

Together, these steps would mitigate the pipeline timeouts/cold-start issues raised by the community:

- Baseline on-demand nodes handle sudden CI/CD bursts without provisioning lag.
- Surge semantics ensure capacity is added before it’s needed during deployments.
- Warm pools guarantee resources for high-priority workloads like parallelized testing.

Would the maintainers be open to a design proposal formalizing this behavior? This could build on existing disruption budgets but add surge semantics.

While third-party solutions (StormForge/CloudPilot/EKS Auto-Mode aka Managed Karpenter) can help, we should prioritize native patterns first. The TTL-to-ReplicaCount concept might complement this by allowing minimum node retention.

Next steps:
a) Consolidate use case requirements
b) Draft enhancement proposal for surge/unavailable parameters
c) Explore warm pool implementation strategies

Would this direction address the core needs? I’m happy to collaborate on formalizing proposals (e.g., for maxSurge/maxUnavailable semantics) or driving use-case consolidation. Let’s ensure we’re aligned with the project’s priorities and maintainers’ vision.

Thanks again for sparking this critical discussion! 🙏

**Comment by ThiyaguE** [comment]
This use-case is much needed one "Manually preprovision a set of nodes before a large event" for us.!

**Comment by cars-matt-b** [comment]
We would love this feature, specifically for what others are saying -- manually provisioning a set of nodes before a large event.

One workaround that comes to mind would be to create a do-nothing deployment with anti-affinity rules restricting one pod to a node. So then in theory, the count of replicas defined would become a pseudo "minimum nodes" count (assuming your pool is able to scale to the size requested).

**Comment by kannon92** [comment]
Putting this here: https://github.com/kubernetes-sigs/kueue/issues/5133

I was thinking about Kueue + Karpenter integration and it seems this is the best issue to cross link here.

I think if ProvisioningRequest is implemented in Karpenter, an admission check would be needed like [CAS](https://kueue.sigs.k8s.io/docs/admission-check-controllers/provisioning/).

What is the status of the HeadRoom proposal?

**Comment by jonathan-innis** [comment]
Adding this here as well: https://docs.google.com/document/d/1bcct-luMPP51YAeUhVuFV7MIXud5wqHsVBDah9WuKeo/edit?tab=t.0#heading=h.kgde2rz09cn8. Something that folks are starting to think about again over in SIG Autoscaling -- based off of the original PodHeadroom proposal

**Comment by jstefankowski** [comment]
Is there a timeline for the buffer API to be available ?    

As others have mentioned in multiple requests for a similar feature, our workflows can be unpredictable, and having additional capacity reserved for heavy load spikes would be especially beneficial for real-time applications.

**Comment by jrleslie** [comment]
I'll also add (because I don't see it mentioned yet) that this will also help with AWS on-demand capacity reservation scenarios. 

With on-demand capacity reservations, instances have essentially been paid for up front based on the on-demand cost of the instance type, so there's no need to have Karpenter scale all the way down to zero for cost savings (because you're billed even if the nodes are off) and will just create delays in node spin up for workloads targeting the capacity reservation.

**Comment by DerekFrank** [comment]
Smaller issue for supporting the K8s Buffer API: https://github.com/kubernetes-sigs/karpenter/issues/2571

**Comment by sumukha-radhakrishna** [comment]
Static Capacity is now supported from v1.8.1 - https://karpenter.sh/docs/concepts/nodepools/#specreplicas

**Comment by frittentheke** [comment]
> Static Capacity is now supported from v1.8.1 - https://karpenter.sh/docs/concepts/nodepools/#specreplicas

While great, this is not "additional" nodes for an otherwise dynamically sized nodepool? 

Even when using a dedicated "over-provisioning" pool with let's say 3 nodes,
pods placed there, render this "buffer" useless as it's not maintained but used up.

Or am I just not understanding "Static capacity" right? @sumukha-radhakrishna

**Comment by frimik** [comment]
> We're considering running karpenter and coredns on Fargate and karpenter then provisioning capacity for everything else.

CoreDNS and Karpenter runs on Fargate in our shop. The rest handled by Karpenter.

**Comment by sumukha-radhakrishna** [comment]
@frittentheke you are correct, static capacity do not maintain additional buffer/headroom, its just a request for spare capacity. We are actively working on implementing the CapacityBuffers API #2571
